Lib/re/grep.py

```diff
@@ -0,0 +1,248 @@
+            yield m.group()
+        return
+
+    i = 0
+    while True:
+        begin = len(line) + 1
+        end = -1
+        best_span = (len(line) + 1, 1)
+        for pat in patterns:
+            m = pat.search(line, i)
+            if m is not None:
+                # Find the longest of the first matches.
+                best_span = min(best_span, (m.begin(), -m.end()))
+                if best_span == (i, -len(line)):
+                    # Matches the rest of line.
+                    break
+        begin = best_span[0]
+        end = -best_span[1]
+        if end < 0:
+            break
+        yield line[begin:end]
+        i = end
+        if begin == end:
+            i += 1
+
+def grep(opts, patterns, file, filename):
+    def print_line(sep, ln, line):
+        if opts.filename:
+            print(filename, end=sep)
+        if opts.line_number:
+            print(ln, end=sep)
+        print(line)
+
+    fast_exit = opts.quiet or opts.filename_only is not None
+    prev_lines = deque(maxlen=opts.before_context)
+    after_lines = 0
+    next_group = None
+    total_count = 0
+    found = False
+    for ln, line in enumerate(file, 1):
+        line = line.removesuffix('\n')
+        if opts.only_matching:
+            for match in findall(patterns, line):
+                print_line(':', ln, match)
+                found = True
+        else:
+            matches = find(patterns, line)
+            if opts.invert_match:
+                matches = not matches
+            if matches:
+                found = True
+                if fast_exit:
+                    break
+                elif opts.count:
+                    total_count += 1
+                else:
+                    if next_group and ln > next_group:
+                        print(opts.group_separator)
+                    for ln2, line2 in enumerate(prev_lines, ln - len(prev_lines)):
+                        print_line('-', ln2, line2)
+                    prev_lines.clear()
+                    print_line(':', ln, line)
+                    after_lines = opts.after_context
+                    next_group = (None if opts.group_separator is None else
+                                  ln + after_lines + opts.before_context + 1)
+            elif not (fast_exit or opts.count):
+                if after_lines:
+                    print_line('-', ln, line)
+                    after_lines -= 1
+                else:
+                    prev_lines.append(line)
+    if fast_exit:
+        if opts.filename_only == found:
+            print(filename)
+    elif opts.count:
+        if opts.filename:
+            print(filename, end=':')
+        print(total_count)
+    return found
+
+def read_from_file(filename):
+    with open(filename, encoding=sys.stdin.encoding,
+              errors=sys.stdin.errors) as f:
+        return [line.removesuffix('\n') for line in f.readlines()]
+
+def main():
+    parser = argparse.ArgumentParser(add_help=False)
```

Comment:
There is a lot of code for one function.  I would put the parser code in a separate function, say 'grep_parser' and replace `opts = parser.parse_args()` below with `opts = grep_parser()`

Reply:
Something went wrong in the quoting.  I only wanted the last line included.